streams/src/main/java/org/apache/kafka/streams/kstream/internals/graph/StreamsGraphNode.java

```diff
@@ -47,10 +48,26 @@ public void setParentNode(final StreamsGraphNode parentNode) {
         this.parentNode = parentNode;
     }
 
-    public Collection<StreamsGraphNode> children() {
+    public Set<StreamsGraphNode> children() {
         return new LinkedHashSet<>(childNodes);
     }
 
+    public void clearChildren() {
+        for (StreamsGraphNode childNode : childNodes) {
+            if (childNode.parentNode != null && childNode.parentNode.equals(this)) {
```

Comment:
Why do we need this check? It seem that `parentNode` cannot be `null`, and that it would be `this` always, too? (assuming one parent node, what is not generic)

Reply:
Since the optimization is pushed out to a 4th PR, this method is removed.  I'll clean up in 4th PR. 

The original idea was to prevent any errors by calling `clearChildren` _*after*_ the children of a given parent node have migrated to another parent.  Probably better to eliminate the checks and document the calling order when updating the graph.